tools/bpf2c/bpf2c_driver.c

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#include <guiddef.h>
+#include <wdm.h>
+#include <wsk.h>
+
+#define NO_CRT
+#include "bpf2c.h"
+
+DRIVER_INITIALIZE DriverEntry;
+DRIVER_UNLOAD DriverUnload;
+RTL_QUERY_REGISTRY_ROUTINE static _bpf2c_query_registry_routine;
+
+static GUID _bpf2c_npi_id = {/* c847aac8-a6f2-4b53-aea3-f4a94b9a80cb */
+                             0xc847aac8,
+                             0xa6f2,
+                             0x4b53,
+                             {0xae, 0xa3, 0xf4, 0xa9, 0x4b, 0x9a, 0x80, 0xcb}};
+static NPI_MODULEID _bpf2c_module_id = {sizeof(_bpf2c_module_id), MIT_GUID, {0}};
+static HANDLE _bpf2c_nmr_client_handle;
+static HANDLE _bpf2c_nmr_provider_handle;
+metadata_table_t ___METADATA_TABLE___;
+
+static NTSTATUS
+_bpf2c_npi_client_attach_provider(
+    _In_ HANDLE nmr_binding_handle,
+    _In_ void* client_context,
+    _In_ const NPI_REGISTRATION_INSTANCE* provider_registration_instance);
+
+static NTSTATUS
+_bpf2c_npi_client_detach_provider(_In_ void* ClientBindingContext);
+
+static const NPI_CLIENT_CHARACTERISTICS _bpf2c_npi_client_characteristics = {
+    0,                                  // Version
+    sizeof(NPI_CLIENT_CHARACTERISTICS), // Length
+    _bpf2c_npi_client_attach_provider,
+    _bpf2c_npi_client_detach_provider,
+    NULL,
+    {0,                                 // Version
+     sizeof(NPI_REGISTRATION_INSTANCE), // Length
+     &_bpf2c_npi_id,
+     &_bpf2c_module_id,
+     0,
+     &___METADATA_TABLE___}};
+
+NTSTATUS
+bpf2c_query_npi_module_id(
+    _In_ const wchar_t* ValueName,
+    unsigned long ValueType,
+    _In_ const void* ValueData,
+    unsigned long ValueLength,
+    _Inout_ void* Context,
+    _Inout_ void* EntryContext)
+{
+    UNREFERENCED_PARAMETER(ValueName);
+    UNREFERENCED_PARAMETER(Context);
+    UNREFERENCED_PARAMETER(EntryContext);
+
+    if (ValueType != REG_BINARY) {
+        return STATUS_INVALID_PARAMETER;
+    }
+    if (ValueLength != sizeof(_bpf2c_module_id.Guid)) {
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    memcpy(&_bpf2c_module_id.Guid, ValueData, ValueLength);
+    return STATUS_SUCCESS;
+}
+
+NTSTATUS
+DriverEntry(_In_ DRIVER_OBJECT* driver_object, _In_ UNICODE_STRING* registry_path)
+{
+    NTSTATUS status;
+    RTL_QUERY_REGISTRY_TABLE query_table[] = {
+        {
+            NULL,                      // Query routine
+            RTL_QUERY_REGISTRY_SUBKEY, // Flags
+            L"Parameters",             // Name
+            NULL,                      // Entry contet
+            REG_NONE,                  // Default type
+            NULL,                      // Default data
+            0,                         // Default length
+        },
+        {
+            bpf2c_query_npi_module_id,   // Query routine
+            RTL_QUERY_REGISTRY_REQUIRED, // Flags
+            L"NpiModuleId",              // Name
```

Comment:
Is the module ID for EC to differentiate between 2 clients? If that is the case, and the skeleton is reading the module ID from registry, I think the following option can simplify the EC design even more?

1. User mode dll will create the service and update a unique NPI provider GUID in the service registry path.
2. Program driver / skeleton will register as NPI provider (same as the extensions today). It will first read the NPI ID from registry and then call NmrRegisterProvider using that NPI ID.
3. User mode will pass the generated NPI ID to EC. EC will use this NPI ID to register as client (same behavior as today when attaching to extensions). This will ensure that attach from EC to skeleton happens synchronously and simplifies the design (same that happens with extensions today).

The above change will ensure that there is 1-1 mapping between EC <--> program driver. Instead, if EC becomes provider, then program driver will attach to EC asynchronously and then EC will need to maintain list of all the clients attached and search for the matching one from the list.

Reply:
I don't think we want to permit an IOCTL that causes EC to bind to an arbitrary NPI ID client. The NPI ID defines the format of the data exchanged between client and provider and by permitting arbitrary NPI IDs we lose type safety.

Reply:
Agree with Alan: one interface ID, multiple module IDs.